In [ ]:
import pandas as pd 

In [ ]:
df = pd.read_excel(r"D:\Project\online+retail\Online Retail.xlsx")

In [ ]:
# display the first 5 rows
df.head() 

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom


In [ ]:
# display the number of rows and columns 
df.shape

(541909, 8)

In [ ]:
# display all column names 
df.columns

Index(['InvoiceNo', 'StockCode', 'Description', 'Quantity', 'InvoiceDate',
       'UnitPrice', 'CustomerID', 'Country'],
      dtype='object')

In [ ]:
# shows the data types and missing-value information
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    541909 non-null  object        
 1   StockCode    541909 non-null  object        
 2   Description  540455 non-null  object        
 3   Quantity     541909 non-null  int64         
 4   InvoiceDate  541909 non-null  datetime64[ns]
 5   UnitPrice    541909 non-null  float64       
 6   CustomerID   406829 non-null  float64       
 7   Country      541909 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 33.1+ MB


### Reprecentation of one each row 
- each row represents one product line within a incoive.
One invoice can contain multiple products, so one invoice can appear in multiple rows. 

### Dataset size 
- number of row: 541909
- number of columns: 8 

### Data Dictionary
- InvoiceNo (object): Unique idetifier for an invoice or transaction
- StockCode (object): Unique identifier for a product 
- Description (object): Name or decription of the product 
- Quantity (int64): Number of product units in the tansaction
- InvoiceData (detertime64): data and time of the transaction
- UniPrice (float64): Price of one product unit
- CustomerID (float64): Identifier for the customer
- Country (object): Country where the customer is located  

In [ ]:
import numpy as np

In [ ]:
# check for missing values 
print(df.isna().sum())

InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64


In [ ]:
# calcuate the percentage of missing value 
print((df.isna().sum()/len(df))*100)

InvoiceNo       0.000000
StockCode       0.000000
Description     0.268311
Quantity        0.000000
InvoiceDate     0.000000
UnitPrice       0.000000
CustomerID     24.926694
Country         0.000000
dtype: float64


In [ ]:
missing_summary = pd.DataFrame({
    "Missing_count": df.isna().sum(),
    "Missing_percentage": (df.isna().sum() / len(df)) * 100
})

missing_summary

,Missing_count,Missing_percentage
InvoiceNo,0,0.000000
StockCode,0,0.000000
Description,1454,0.268311
Quantity,0,0.000000
InvoiceDate,0,0.000000
UnitPrice,0,0.000000
CustomerID,135080,24.926694
Country,0,0.000000


In [ ]:
df.duplicated().sum()

np.int64(5268)

In [ ]:
# check for duplicate rows 
duplicate_rows = df[df.duplicated(keep=False)]
duplicate_rows.head(10)

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
485,536409,22111,SCOTTIE DOG HOT WATER BOTTLE,1,2010-12-01 11:45:00,4.95,17908.0,United Kingdom
489,536409,22866,HAND WARMER SCOTTY DOG DESIGN,1,2010-12-01 11:45:00,2.10,17908.0,United Kingdom
494,536409,21866,UNION JACK FLAG LUGGAGE TAG,1,2010-12-01 11:45:00,1.25,17908.0,United Kingdom
517,536409,21866,UNION JACK FLAG LUGGAGE TAG,1,2010-12-01 11:45:00,1.25,17908.0,United Kingdom
521,536409,22900,SET 2 TEA TOWELS I LOVE LONDON,1,2010-12-01 11:45:00,2.95,17908.0,United Kingdom
527,536409,22866,HAND WARMER SCOTTY DOG DESIGN,1,2010-12-01 11:45:00,2.10,17908.0,United Kingdom
537,536409,22900,SET 2 TEA TOWELS I LOVE LONDON,1,2010-12-01 11:45:00,2.95,17908.0,United Kingdom
539,536409,22111,SCOTTIE DOG HOT WATER BOTTLE,1,2010-12-01 11:45:00,4.95,17908.0,United Kingdom
548,536412,22327,ROUND SNACK BOXES SET OF 4 SKULLS,1,2010-12-01 11:49:00,2.95,17920.0,United Kingdom
555,536412,22327,ROUND SNACK BOXES SET OF 4 SKULLS,1,2010-12-01 11:49:00,2.95,17920.0,United Kingdom


In [ ]:
duplicate_rows.sort_values(
    by=["InvoiceNo", "StockCode", "InvoiceDate"]
).head(20)

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
494,536409,21866,UNION JACK FLAG LUGGAGE TAG,1,2010-12-01 11:45:00,1.25,17908.0,United Kingdom
517,536409,21866,UNION JACK FLAG LUGGAGE TAG,1,2010-12-01 11:45:00,1.25,17908.0,United Kingdom
485,536409,22111,SCOTTIE DOG HOT WATER BOTTLE,1,2010-12-01 11:45:00,4.95,17908.0,United Kingdom
539,536409,22111,SCOTTIE DOG HOT WATER BOTTLE,1,2010-12-01 11:45:00,4.95,17908.0,United Kingdom
489,536409,22866,HAND WARMER SCOTTY DOG DESIGN,1,2010-12-01 11:45:00,2.10,17908.0,United Kingdom
527,536409,22866,HAND WARMER SCOTTY DOG DESIGN,1,2010-12-01 11:45:00,2.10,17908.0,United Kingdom
521,536409,22900,SET 2 TEA TOWELS I LOVE LONDON,1,2010-12-01 11:45:00,2.95,17908.0,United Kingdom
537,536409,22900,SET 2 TEA TOWELS I LOVE LONDON,1,2010-12-01 11:45:00,2.95,17908.0,United Kingdom
565,536412,21448,12 DAISY PEGS IN WOOD BOX,2,2010-12-01 11:49:00,1.65,17920.0,United Kingdom
578,536412,21448,12 DAISY PEGS IN WOOD BOX,1,2010-12-01 11:49:00,1.65,17920.0,United Kingdom


In [ ]:
duplicate_percentage = df.duplicated().sum() / len(df) * 100
duplicate_percentage

np.float64(0.9721189350979592)

In [ ]:
# Negative quantities (cancellation or return)
negative_quantity = (df["Quantity"] < 0).sum()
negative_quantity

np.int64(10624)

In [ ]:
df[df['Quantity']<0].head(10)

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
141,C536379,D,Discount,-1,2010-12-01 09:41:00,27.50,14527.0,United Kingdom
154,C536383,35004C,SET OF 3 COLOURED FLYING DUCKS,-1,2010-12-01 09:49:00,4.65,15311.0,United Kingdom
235,C536391,22556,PLASTERS IN TIN CIRCUS PARADE,-12,2010-12-01 10:24:00,1.65,17548.0,United Kingdom
236,C536391,21984,PACK OF 12 PINK PAISLEY TISSUES,-24,2010-12-01 10:24:00,0.29,17548.0,United Kingdom
237,C536391,21983,PACK OF 12 BLUE PAISLEY TISSUES,-24,2010-12-01 10:24:00,0.29,17548.0,United Kingdom
238,C536391,21980,PACK OF 12 RED RETROSPOT TISSUES,-24,2010-12-01 10:24:00,0.29,17548.0,United Kingdom
239,C536391,21484,CHICK GREY HOT WATER BOTTLE,-12,2010-12-01 10:24:00,3.45,17548.0,United Kingdom
240,C536391,22557,PLASTERS IN TIN VINTAGE PAISLEY,-12,2010-12-01 10:24:00,1.65,17548.0,United Kingdom
241,C536391,22553,PLASTERS IN TIN SKULLS,-24,2010-12-01 10:24:00,1.65,17548.0,United Kingdom
939,C536506,22960,JAM MAKING SET WITH JARS,-6,2010-12-01 12:38:00,4.25,17897.0,United Kingdom


In [ ]:
# Zero quantities
zero_quantity = (df['Quantity']==0).sum()
zero_quantity

np.int64(0)

In [ ]:
# Negative Prices 
negative_price=(df['UnitPrice']<0).sum()
negative_price

np.int64(2)

In [ ]:
df[df["UnitPrice"] < 0].head(10)

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
299983,A563186,B,Adjust bad debt,1,2011-08-12 14:51:00,-11062.06,NaN,United Kingdom
299984,A563187,B,Adjust bad debt,1,2011-08-12 14:52:00,-11062.06,NaN,United Kingdom


In [ ]:
#zero price = free product 
zero_price = (df["UnitPrice"] == 0).sum()
zero_price

np.int64(2515)

In [ ]:
df[df["UnitPrice"] == 0].head(10)

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
622,536414,22139,NaN,56,2010-12-01 11:52:00,0.0,NaN,United Kingdom
1970,536545,21134,NaN,1,2010-12-01 14:32:00,0.0,NaN,United Kingdom
1971,536546,22145,NaN,1,2010-12-01 14:33:00,0.0,NaN,United Kingdom
1972,536547,37509,NaN,1,2010-12-01 14:33:00,0.0,NaN,United Kingdom
1987,536549,85226A,NaN,1,2010-12-01 14:34:00,0.0,NaN,United Kingdom
1988,536550,85044,NaN,1,2010-12-01 14:34:00,0.0,NaN,United Kingdom
2024,536552,20950,NaN,1,2010-12-01 14:34:00,0.0,NaN,United Kingdom
2025,536553,37461,NaN,3,2010-12-01 14:35:00,0.0,NaN,United Kingdom
2026,536554,84670,NaN,23,2010-12-01 14:35:00,0.0,NaN,United Kingdom
2406,536589,21777,NaN,-10,2010-12-01 16:50:00,0.0,NaN,United Kingdom


In [ ]:
# cancellation 
cancelled_invoice = (
    df["InvoiceNo"]
    .astype(str)
    .str.startswith("C")
    .sum()
)

cancelled_invoice

np.int64(9288)

In [ ]:
df[
    df["InvoiceNo"].astype(str).str.startswith("C")
].head(10)

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
141,C536379,D,Discount,-1,2010-12-01 09:41:00,27.50,14527.0,United Kingdom
154,C536383,35004C,SET OF 3 COLOURED FLYING DUCKS,-1,2010-12-01 09:49:00,4.65,15311.0,United Kingdom
235,C536391,22556,PLASTERS IN TIN CIRCUS PARADE,-12,2010-12-01 10:24:00,1.65,17548.0,United Kingdom
236,C536391,21984,PACK OF 12 PINK PAISLEY TISSUES,-24,2010-12-01 10:24:00,0.29,17548.0,United Kingdom
237,C536391,21983,PACK OF 12 BLUE PAISLEY TISSUES,-24,2010-12-01 10:24:00,0.29,17548.0,United Kingdom
238,C536391,21980,PACK OF 12 RED RETROSPOT TISSUES,-24,2010-12-01 10:24:00,0.29,17548.0,United Kingdom
239,C536391,21484,CHICK GREY HOT WATER BOTTLE,-12,2010-12-01 10:24:00,3.45,17548.0,United Kingdom
240,C536391,22557,PLASTERS IN TIN VINTAGE PAISLEY,-12,2010-12-01 10:24:00,1.65,17548.0,United Kingdom
241,C536391,22553,PLASTERS IN TIN SKULLS,-24,2010-12-01 10:24:00,1.65,17548.0,United Kingdom
939,C536506,22960,JAM MAKING SET WITH JARS,-6,2010-12-01 12:38:00,4.25,17897.0,United Kingdom


In [ ]:
quality_counts = pd.Series({
    "Exact duplicate rows": df.duplicated().sum(),
    "Negative quantity rows": negative_quantity,
    "Zero quantity rows": zero_quantity,
    "Negative price rows": negative_price,
    "Zero price rows": zero_price,
    "Cancelled invoice rows": cancelled_invoice
})

quality_counts

Exact duplicate rows       5268
Negative quantity rows    10624
Zero quantity rows            0
Negative price rows           2
Zero price rows            2515
Cancelled invoice rows     9288
dtype: int64

- Duplicate rows: 
- Negative quantity rows: retruned products, cancelled orders, damaged-product, incentory corrections...
- Zero quantity: there are no row for this 
- Negative price: data errors, accounting adjustments 
- Zero price rows: free staffs, prometional products, missing price info, data entry problems 
- Cancelled invoice 

### Assume that relationship between negative quantity and cancelled invoice

- Negative quantity rows are 10624 
- Cancelled invoice rows are 9288

- Assume that every cancelled row has a negative quantity
(10624-2988 = 1336) -> 1336 negatice quantity rows do not begin with C

In [ ]:
# check for the relationship between negative quantity and cancelled invoice
cancelled_mask = df["InvoiceNo"].astype(str).str.startswith("C")
negative_quantity_mask = df["Quantity"] <0

(cancelled_mask & negative_quantity_mask).sum()

np.int64(9288)

In [ ]:
(negative_quantity_mask & ~cancelled_mask).sum()

np.int64(1336)

In [ ]:
df[negative_quantity_mask & ~cancelled_mask].head(10)

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
2406,536589,21777,NaN,-10,2010-12-01 16:50:00,0.0,NaN,United Kingdom
4347,536764,84952C,NaN,-38,2010-12-02 14:42:00,0.0,NaN,United Kingdom
7188,536996,22712,NaN,-20,2010-12-03 15:30:00,0.0,NaN,United Kingdom
7189,536997,22028,NaN,-20,2010-12-03 15:30:00,0.0,NaN,United Kingdom
7190,536998,85067,NaN,-6,2010-12-03 15:30:00,0.0,NaN,United Kingdom
7192,537000,21414,NaN,-22,2010-12-03 15:32:00,0.0,NaN,United Kingdom
7193,537001,21653,NaN,-6,2010-12-03 15:33:00,0.0,NaN,United Kingdom
7195,537003,85126,NaN,-2,2010-12-03 15:33:00,0.0,NaN,United Kingdom
7196,537004,21814,NaN,-30,2010-12-03 15:34:00,0.0,NaN,United Kingdom
7197,537005,21692,NaN,-70,2010-12-03 15:35:00,0.0,NaN,United Kingdom


So all of cancellation invioce has negative quantity

# Clean the data

In [ ]:
# copy of the original data
clean_df = df.copy()

In [ ]:
# starting rows 
starting_rows = len(clean_df)
starting_rows

541909

In [ ]:
# create a cleaning log for recording 
cleaning_log = []

cleaning_log.append({
    "Step":"Initial data",
    "Rows_Before":len(clean_df),
    "Rows_After": len(clean_df),
    "Rows_Removed": 0
})

In [ ]:
# Remove exact duplicates
rows_before_du = len(clean_df)

In [ ]:
clean_df = clean_df.drop_duplicates().copy()

In [ ]:
rows_after_du = len(clean_df)

In [ ]:
rows_removed_du = (rows_before_du - rows_after_du)

rows_removed_du

5268

In [ ]:
cleaning_log.append({
    "Step":"Removed exact duplicates",
    "Rows_Before":rows_before_du,
    "Rows_After": rows_after_du,
    "Rows_Removed": rows_removed_du
})

In [ ]:
cleaning_log

[{'Step': 'Initial data',
  'Rows_Before': 541909,
  'Rows_After': 541909,
  'Rows_Removed': 0},
 {'Step': 'Removed exact duplicates',
  'Rows_Before': 541909,
  'Rows_After': 536641,
  'Rows_Removed': 5268}]

In [ ]:
clean_df.shape

(536641, 8)

In [ ]:
#convert the InvoiceNo to string (text)
clean_df["InvoiceNo"]=(clean_df["InvoiceNo"].astype("string"))

In [ ]:
#convert the StockCode to string (text)
clean_df["StockCode"]=(clean_df["StockCode"].astype("string"))

In [ ]:
# Convert the CustomerID from decimal to integer, then text
clean_df["CustomerID"] = (
    clean_df["CustomerID"].astype("Int64").astype("string")
)

In [ ]:
# Ensure the data is a datetime 
clean_df["InvoiceDate"] = pd.to_datetime(
    clean_df["InvoiceDate"],
    errors="coerce"
)

In [ ]:
#cehck the data types of the columns
clean_df.dtypes

InvoiceNo      string[python]
StockCode      string[python]
Description            object
Quantity                int64
InvoiceDate    datetime64[ns]
UnitPrice             float64
CustomerID     string[python]
Country                object
dtype: object

In [ ]:
# organise the space in the "description" column
clean_df["Description"] = (
    clean_df["Description"]
    .astype("string")
    .str.replace(r"\s+", " ", regex=True)
    .str.strip()
)

In [ ]:
# organise the space in the "country" column
clean_df["Country"] = (
    clean_df["Country"]
    .astype("string")
    .str.replace(r"\s+", " ", regex=True)
    .str.strip()
)

## Create transaction status 

In [ ]:
clean_df["TransactionStatus"] = "Completed Sales"

In [ ]:
# identify missing values in the InvoiceNo column
cancelled_mask = (
    clean_df["InvoiceNo"].str.startswith("C", na=False)
)

In [ ]:
# identify cancelled transactions for updating the TransactionStatus column
clean_df.loc[
    cancelled_mask,
    "TransactionStatus"
] = "Cancelled"

In [ ]:
# identify negative quantities without C
stock_adjustment_mask = (
    (clean_df["Quantity"]<0) & ~cancelled_mask
)

In [ ]:
# identify stock adjustment transactions for updating the TransactionStatus column
clean_df.loc[
    stock_adjustment_mask,
    "TransactionStatus"
] = "Stock Adjustment"

In [ ]:
#identify zero price rows 
zero_price_mask = clean_df["UnitPrice"] == 0

In [ ]:
#identify zero price transactions for updating the TransactionStatus column
clean_df.loc[
    zero_price_mask,
    "TransactionStatus"
] = "Price Adjustment"

In [ ]:
#identify negative price rows
negative_price_mask = clean_df["UnitPrice"] < 0

In [ ]:
# identify negative price transactions for updating the TransactionStatus column
clean_df.loc[
    negative_price_mask,
    "TransactionStatus"
] = "Price Adjustment"

In [ ]:
# cehck the TransactionStatus 
clean_df["TransactionStatus"].value_counts(
    dropna=False
)

TransactionStatus
Completed Sales     524878
Cancelled             9251
Price Adjustment      2512
Name: count, dtype: int64

In [ ]:
# check the examples 
clean_df[
    [
        "InvoiceNo",
        "Description",
        "Quantity",
        "UnitPrice",
        "TransactionStatus"
    ]
].head(20)

,InvoiceNo,Description,Quantity,UnitPrice,TransactionStatus
0,536365,WHITE HANGING HEART T-LIGHT HOLDER,6,2.55,Completed Sales
1,536365,WHITE METAL LANTERN,6,3.39,Completed Sales
2,536365,CREAM CUPID HEARTS COAT HANGER,8,2.75,Completed Sales
3,536365,KNITTED UNION FLAG HOT WATER BOTTLE,6,3.39,Completed Sales
4,536365,RED WOOLLY HOTTIE WHITE HEART.,6,3.39,Completed Sales
5,536365,SET 7 BABUSHKA NESTING BOXES,2,7.65,Completed Sales
6,536365,GLASS STAR FROSTED T-LIGHT HOLDER,6,4.25,Completed Sales
7,536366,HAND WARMER UNION JACK,6,1.85,Completed Sales
8,536366,HAND WARMER RED POLKA DOT,6,1.85,Completed Sales
9,536367,ASSORTED COLOUR BIRD ORNAMENT,32,1.69,Completed Sales


In [ ]:
cleaning_log.append({
    "Step": "Corrected types, standardised text and labelled status",
    "Rows_Before": len(clean_df),
    "Rows_After": len(clean_df),
    "Rows_Removed": 0
})

In [ ]:
cleaning_log_df = pd.DataFrame(cleaning_log)

cleaning_log_df

,Step,Rows_Before,Rows_After,Rows_Removed
0,Initial data,541909,541909,0
1,Removed exact duplicates,541909,536641,5268
2,"Corrected types, standardised text and labelle...",536641,536641,0


In [ ]:
clean_df.shape

(536641, 9)

In [ ]:
clean_df.dtypes

InvoiceNo            string[python]
StockCode            string[python]
Description          string[python]
Quantity                      int64
InvoiceDate          datetime64[ns]
UnitPrice                   float64
CustomerID           string[python]
Country              string[python]
TransactionStatus            object
dtype: object

In [ ]:
clean_df.isna().sum()

InvoiceNo                 0
StockCode                 0
Description            1454
Quantity                  0
InvoiceDate               0
UnitPrice                 0
CustomerID           135037
Country                   0
TransactionStatus         0
dtype: int64

In [ ]:
clean_df.duplicated().sum()

np.int64(0)

In [ ]:
clean_df["TransactionStatus"].value_counts()

TransactionStatus
Completed Sales     524878
Cancelled             9251
Price Adjustment      2512
Name: count, dtype: int64

# Find the useful features like revenue and time 

In [ ]:
clean_df["Revenue"] = (
    clean_df["Quantity"] * clean_df["UnitPrice"]
)

In [ ]:
clean_df[
    [
        "InvoiceNo",
        "Quantity",
        "UnitPrice",
        "Revenue",
        "TransactionStatus"
    ]
].head(10)

,InvoiceNo,Quantity,UnitPrice,Revenue,TransactionStatus
0,536365,6,2.55,15.30,Completed Sales
1,536365,6,3.39,20.34,Completed Sales
2,536365,8,2.75,22.00,Completed Sales
3,536365,6,3.39,20.34,Completed Sales
4,536365,6,3.39,20.34,Completed Sales
5,536365,2,7.65,15.30,Completed Sales
6,536365,6,4.25,25.50,Completed Sales
7,536366,6,1.85,11.10,Completed Sales
8,536366,6,1.85,11.10,Completed Sales
9,536367,32,1.69,54.08,Completed Sales


In [ ]:
clean_df.shape

(536641, 10)

In [ ]:
# check for the revenue error 
revenue_error = (
    clean_df["Quantity"] * clean_df["UnitPrice"] != clean_df["Revenue"]
).sum()

revenue_error

np.int64(0)

In [ ]:
# check for the missing revenue values 
clean_df["Revenue"].isna().sum()

np.int64(0)

In [ ]:
# how cancellations affect the revenue 
cancelled_revenue = clean_df[
    clean_df["InvoiceNo"].str.startswith(
        "C", na=False
    )
]

cancelled_revenue[
    ["InvoiceNo", "Quantity", "UnitPrice", "Revenue", "TransactionStatus"]
].head(10)


,InvoiceNo,Quantity,UnitPrice,Revenue,TransactionStatus
141,C536379,-1,27.50,-27.50,Cancelled
154,C536383,-1,4.65,-4.65,Cancelled
235,C536391,-12,1.65,-19.80,Cancelled
236,C536391,-24,0.29,-6.96,Cancelled
237,C536391,-24,0.29,-6.96,Cancelled
238,C536391,-24,0.29,-6.96,Cancelled
239,C536391,-12,3.45,-41.40,Cancelled
240,C536391,-12,1.65,-19.80,Cancelled
241,C536391,-24,1.65,-39.60,Cancelled
939,C536506,-6,4.25,-25.50,Cancelled


In [ ]:
# calculate the revenue for each transactionStatus 
revenue_by_status = (
    clean_df
    .groupby("TransactionStatus")["Revenue"]
    .agg(["count","sum"])
)

revenue_by_status

,count,sum
TransactionStatus,,
Cancelled,9251,-8.939797e+05
Completed Sales,524878,1.064211e+07
Price Adjustment,2512,-2.212412e+04


In [ ]:
# calculate the net recorded revenue
net_recorded_revenue = clean_df["Revenue"].sum()

net_recorded_revenue

np.float64(9726006.953999998)

In [ ]:
#check for date feature 
clean_df["InvoiceDate"].dtype

dtype('<M8[ns]')

In [ ]:
clean_df["invoiceData"] = pd.to_datetime(
    clean_df["InvoiceDate"],
    errors = "coerce"
)

In [ ]:
# Year and Month
clean_df["YearMonth"] = (
    clean_df["InvoiceDate"]
    .dt.to_period("M")
    .astype("string")
)

In [ ]:
# Weekday
clean_df["Weekday"] = (
    clean_df["InvoiceDate"].dt.day_name()
)

In [ ]:
# Hour
clean_df["Hour"] =(
    clean_df["InvoiceDate"].dt.hour
)

In [ ]:
# Date
clean_df["Date"]=(
    clean_df["InvoiceDate"].dt.date
)

In [ ]:
clean_df[
    ["InvoiceDate", "YearMonth","Weekday", "Hour", "Date"]
].head(10)

,InvoiceDate,YearMonth,Weekday,Hour,Date
0,2010-12-01 08:26:00,2010-12,Wednesday,8,2010-12-01
1,2010-12-01 08:26:00,2010-12,Wednesday,8,2010-12-01
2,2010-12-01 08:26:00,2010-12,Wednesday,8,2010-12-01
3,2010-12-01 08:26:00,2010-12,Wednesday,8,2010-12-01
4,2010-12-01 08:26:00,2010-12,Wednesday,8,2010-12-01
5,2010-12-01 08:26:00,2010-12,Wednesday,8,2010-12-01
6,2010-12-01 08:26:00,2010-12,Wednesday,8,2010-12-01
7,2010-12-01 08:28:00,2010-12,Wednesday,8,2010-12-01
8,2010-12-01 08:28:00,2010-12,Wednesday,8,2010-12-01
9,2010-12-01 08:34:00,2010-12,Wednesday,8,2010-12-01


In [ ]:
# calculate info for each invoice
order_summary = (
    clean_df
    .groupby("InvoiceNo")
    .agg(
        OrderRevenue=("Revenue", "sum"),
        DistinctProducts=("StockCode", "nunique"),
        TransactionRows=("StockCode", "size")
    )
    .reset_index()
)

In [ ]:
order_summary.head(10)

,InvoiceNo,OrderRevenue,DistinctProducts,TransactionRows
0,536365,139.12,7,7
1,536366,22.20,2,2
2,536367,278.73,12,12
3,536368,70.05,4,4
4,536369,17.85,1,1
5,536370,855.86,20,20
6,536371,204.00,1,1
7,536372,22.20,2,2
8,536373,259.86,16,16
9,536374,350.40,1,1


In [ ]:
# select a sample invoice 
sample_invoice = order_summary.iloc[0]["InvoiceNo"]

sample_invoice

'536365'

In [ ]:
clean_df[
    clean_df["InvoiceNo"] == sample_invoice
][
    [
        "InvoiceNo",
        "StockCode",
        "Quantity",
        "UnitPrice",
        "Revenue"
    ]
]

,InvoiceNo,StockCode,Quantity,UnitPrice,Revenue
0,536365,85123A,6,2.55,15.30
1,536365,71053,6,3.39,20.34
2,536365,84406B,8,2.75,22.00
3,536365,84029G,6,3.39,20.34
4,536365,84029E,6,3.39,20.34
5,536365,22752,2,7.65,15.30
6,536365,21730,6,4.25,25.50


In [ ]:
order_summary[
    order_summary["InvoiceNo"] == sample_invoice
]

,InvoiceNo,OrderRevenue,DistinctProducts,TransactionRows
0,536365,139.12,7,7


In [ ]:
clean_df.shape

(536641, 15)

In [ ]:
clean_df.columns

Index(['InvoiceNo', 'StockCode', 'Description', 'Quantity', 'InvoiceDate',
       'UnitPrice', 'CustomerID', 'Country', 'TransactionStatus', 'Revenue',
       'invoiceData', 'YearMonth', 'Weekday', 'Hour', 'Date'],
      dtype='object')

In [ ]:
clean_df.isna().sum()

InvoiceNo                 0
StockCode                 0
Description            1454
Quantity                  0
InvoiceDate               0
UnitPrice                 0
CustomerID           135037
Country                   0
TransactionStatus         0
Revenue                   0
invoiceData               0
YearMonth                 0
Weekday                   0
Hour                      0
Date                      0
dtype: int64

In [ ]:
clean_df.duplicated().sum()

np.int64(0)

In [ ]:
clean_df.dtypes

InvoiceNo            string[python]
StockCode            string[python]
Description          string[python]
Quantity                      int64
InvoiceDate          datetime64[ns]
UnitPrice                   float64
CustomerID           string[python]
Country              string[python]
TransactionStatus            object
Revenue                     float64
invoiceData          datetime64[ns]
YearMonth            string[python]
Weekday                      object
Hour                          int32
Date                         object
dtype: object

In [ ]:
clean_df = clean_df.drop(
    columns=["invoiceDate"]
)

In [ ]:
clean_df.columns

In [ ]:
unique_invoices = clean_df["InvoiceNo"].nunique()

unique_invoices

25900

In [ ]:
unique_customers = clean_df["CustomerID"].nunique()

unique_customers

4372

In [ ]:
unique_countries = clean_df["Country"].nunique()

unique_countries

38

In [ ]:
sorted(clean_df["Country"].dropna().unique())

['Australia',
 'Austria',
 'Bahrain',
 'Belgium',
 'Brazil',
 'Canada',
 'Channel Islands',
 'Cyprus',
 'Czech Republic',
 'Denmark',
 'EIRE',
 'European Community',
 'Finland',
 'France',
 'Germany',
 'Greece',
 'Hong Kong',
 'Iceland',
 'Israel',
 'Italy',
 'Japan',
 'Lebanon',
 'Lithuania',
 'Malta',
 'Netherlands',
 'Norway',
 'Poland',
 'Portugal',
 'RSA',
 'Saudi Arabia',
 'Singapore',
 'Spain',
 'Sweden',
 'Switzerland',
 'USA',
 'United Arab Emirates',
 'United Kingdom',
 'Unspecified']

In [ ]:
clean_df["TransactionStatus"].value_counts(
    dropna=False
)

TransactionStatus
Completed Sales     524878
Cancelled             9251
Price Adjustment      2512
Name: count, dtype: int64

In [ ]:
clean_df["TransactionStatus"].isna().sum()

np.int64(0)

In [ ]:
validation_summary = pd.Series({
    "Total rows": len(clean_df),
    "Total columns": clean_df.shape[1],
    "Exact duplicates": clean_df.duplicated().sum(),
    "Unique invoices": clean_df["InvoiceNo"].nunique(),
    "Unique customers": clean_df["CustomerID"].nunique(),
    "Unique countries": clean_df["Country"].nunique(),
    "Earliest date": clean_df["InvoiceDate"].min(),
    "Latest date": clean_df["InvoiceDate"].max()
})

validation_summary

Total rows                       536641
Total columns                        15
Exact duplicates                      0
Unique invoices                   25900
Unique customers                   4372
Unique countries                     38
Earliest date       2010-12-01 08:26:00
Latest date         2011-12-09 12:50:00
dtype: object

## Create the export datasets

In [ ]:
clean_transaction = clean_df.copy()

### Exports completed sales datasets

In [ ]:
clean_df["TransactionStatus"].unique()

array(['Completed Sales', 'Cancelled', 'Price Adjustment'], dtype=object)

In [ ]:
completed_sales = clean_df[
    clean_df["TransactionStatus"] == "Completed Sales"
].copy()

In [ ]:
completed_sales.shape

(524878, 15)

In [ ]:
completed_sales[
    "TransactionStatus"
].value_counts()

TransactionStatus
Completed Sales    524878
Name: count, dtype: int64

### Export cancellations datasets

In [ ]:
cancelled = clean_df[clean_df["TransactionStatus"]=="Cancelled"].copy()

In [ ]:
cancelled.shape

(9251, 15)

In [ ]:
cancelled[
    "TransactionStatus"
].value_counts()

TransactionStatus
Cancelled    9251
Name: count, dtype: int64

In [ ]:
from pathlib import Path

In [ ]:
output_folder = Path(
    r"D:\Project\online+retail\processed"
)

output_folder.mkdir(
    parents=True,
    exist_ok=True
)

In [ ]:
clean_transaction.to_csv(
    output_folder / "clean_transactions.csv",
    index=False
)

In [ ]:
completed_sales.to_csv(
    output_folder / "completed_sales.csv",
    index=False
)

In [ ]:
cancelled.to_csv(
    output_folder / "cancellations.csv",
    index=False
)

In [ ]:
list(output_folder.glob("*.csv"))

In [ ]:
test_clean = pd.read_csv(
    output_folder / "clean_transactions.csv"
)

test_completed = pd.read_csv(
    output_folder / "completed_sales.csv"
)

test_cancellations = pd.read_csv(
    output_folder / "cancellations.csv"
)

C:\Users\yyjj0\AppData\Local\Temp\ipykernel_23980\3879385134.py:5: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.
  test_completed = pd.read_csv(


In [ ]:
test_clean = pd.read_csv(
    output_folder / "clean_transactions.csv"
)

test_completed = pd.read_csv(
    output_folder / "completed_sales.csv"
)

test_cancellations = pd.read_csv(
    output_folder / "cancellations.csv"
)

C:\Users\yyjj0\AppData\Local\Temp\ipykernel_23980\3879385134.py:5: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.
  test_completed = pd.read_csv(
